### PR TITLE
add gemm and batchmatmul, used intel mkl cblas_sgemm, oneDNN dnnl_sge…

### PR DESCRIPTION
**Pull request by simon2**
### added files in kernels/gemm
I added folder _mkl_ and _oneDNN_ in `kernels/gemm`, both of these two folders contains a **main.cpp**. 
the one in _mkl_ is a normal gemm implemented by intel MKL `cblas_sgemm`, 
and the one in _oneDNN_ is a normal gemm implemented by `dnnl_sgemm` function of intel oneDNN.

### added files in kernels/batchmatmul
I added **args.hpp** for data initialization for bmm at `kernels/batchmatmul`.
I also added folder _mkl_ and _oneDNN_ in `kernels/batchmatmul`.
**main.cpp** in `batchmatmul/mkl` is an intel MKL `cblas_sgemm` inside a OpenMP for-loop implementation.
**dnnl.cpp** in `batchmatmul/oneDNN` is an intel oneDNN `dnnl_sgemm` inside a OpenMP for-loop implementation.
**matmul.cpp** in `batchmatmul/oneDNN` is an intel oneDNN `matmul primitive` implementation, which do not have a outside omp-for-loop.
**example_utils.hpp** in `batchmatmul/oneDNN` is just used for `matmul.cpp`, which is directly copied from the example code that oneDNN give us.

### CODE REVIEW

File: benchmarker/kernels/gemm/mkl/Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#icpc -std=c++11 -fopenmp -o main main.cpp -mkl
+.PHONY: run
+
+CXXFLAGS=-Wall -O3  -fopenmp --std=c++14 -msse3  -ftree-vectorize -ftree-vectorizer-verbose=3 -fopt-info-vec
+
+CFLAGS=-Wall -O3  -fopenmp  -ftree-vectorize -ftree-vectorizer-verbose=3 -fopt-info-vec
```
Comment (vatai) [inline benchmarker/kernels/gemm/mkl/Makefile:6]: You don't seem to be using `CFLAGS` anywhere. If so, delete it.

File: benchmarker/kernels/gemm/mkl/Makefile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#icpc -std=c++11 -fopenmp -o main main.cpp -mkl
+.PHONY: run
+
+CXXFLAGS=-Wall -O3  -fopenmp --std=c++14 -msse3  -ftree-vectorize -ftree-vectorizer-verbose=3 -fopt-info-vec
```
Comment (vatai) [inline benchmarker/kernels/gemm/mkl/Makefile:4]: the `-mkl` should be included with `CXXFLAGS`

File: benchmarker/kernels/gemm/mkl/Makefile [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+CFLAGS=-Wall -O3  -fopenmp  -ftree-vectorize -ftree-vectorizer-verbose=3 -fopt-info-vec
+
+
+all: main
+
+main: main.cpp
+	# mkl only works for intel compiler
+	icpc -o $@ $^ $(CXXFLAGS) -mkl
```
Comment (vatai) [inline benchmarker/kernels/gemm/mkl/Makefile:13]: If you define `CXX=icpc` near the top, e.g. right before `CXXFLAGS` then you can remove these two line (12 and 13), make will know what to do!

**Review by vatai — CHANGES_REQUESTED** [review]

File: benchmarker/frameworks/do_mkl.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        super().__init__(params, remaining_args)
+        assert self.params["problem"]["precision"] in ["FP32", "mixed"]
+
+    def run(self):
+        if "nb_gpus" in self.params:
+            if self.params["nb_gpus"] > 0:
+                raise Exception("cblas does not work on GPU")
+        # TODO(Alex): this does not work inless the binaries are copied to site_packages
+        dirname = os.path.dirname(os.path.realpath(__file__))
+        path_binary = os.path.join(dirname, "../kernels/gemm/mkl/main")
```
Comment (vatai) [inline benchmarker/frameworks/do_mkl.py:19]: This path seems to be wrong... actually this might be a bigger problem...

**Review by vatai — CHANGES_REQUESTED** [review]

File: benchmarker/frameworks/do_mkl.py (hunk 4 above)
Comment (vatai) [reply benchmarker/frameworks/do_mkl.py:19]: Try this (or something similar) and check that it works

```python
path_binary = os.path.join(dirname, "../kernels", self.params["problem"], self.params["framework"], "main")
```

Comment (simon2) [reply benchmarker/frameworks/do_mkl.py:19]: Thank you, but `self.params["problem"] `is a `dict`, so I added get function after it.
`path_binary = os.path.join(dirname, "../kernels", self.params["problem"].get("name"), self.params["framework"], "main")`

**Pull request merged**